File: frontend/src/lib/taskTypePaths.test.ts
import { describe, expect, it } from 'vitest'
import type { TaskType } from './api'
import {
  buildTaskTypeSuggestions,
  canonicalizeTaskTypePathInput,
  createAncestorHint,
  filterTaskTypesByQuery,
  formatTaskTypePathParts,
  groupTaskTypesByRoot,
  pathDepth,
  rankTaskTypes,
  taskTypeRootSegment,
} from './taskTypePaths'

const rows: TaskType[] = [
  { id: 1, name: 'coding', created_at: '', updated_at: '' },
  { id: 2, name: 'coding/ai', created_at: '', updated_at: '' },
  { id: 3, name: 'exercise/cardio', created_at: '', updated_at: '' },
]

describe('taskTypePaths', () => {
  it('canonicalizes slash-delimited input', () => {
    expect(canonicalizeTaskTypePathInput(' Coding / AI ')).toBe('coding/ai')
  })

  it('drops empty segments and trailing slashes', () => {
    expect(canonicalizeTaskTypePathInput('coding//ai')).toBe('coding/ai')
    expect(canonicalizeTaskTypePathInput('coding/')).toBe('coding')
    expect(canonicalizeTaskTypePathInput('coding//ai/')).toBe('coding/ai')
  })

  it('collapses inner whitespace in segments', () => {
    expect(canonicalizeTaskTypePathInput('Deep   Work / Writing')).toBe('deep work/writing')
  })

  it('returns null when nothing survives', () => {
    expect(canonicalizeTaskTypePathInput('')).toBeNull()
    expect(canonicalizeTaskTypePathInput('   ')).toBeNull()
    expect(canonicalizeTaskTypePathInput('///')).toBeNull()
  })

  it('splits display into ancestor and leaf parts', () => {
    expect(formatTaskTypePathParts('coding/ai/agents')).toEqual({
      ancestorsLabel: 'coding / ai',
      leafLabel: 'agents',
      fullLabel: 'coding/ai/agents',
    })
  })

  it('offers a create row only when the canonical path is missing', () => {
    const suggestions = buildTaskTypeSuggestions(rows, 'coding/personal')
    expect(suggestions.createPath).toBe('coding/personal')
    expect(suggestions.rows.map((row) => row.name)).toContain('coding')
  })

  it('offers create for a repaired new path', () => {
    expect(buildTaskTypeSuggestions(rows, 'coding//personal/').createPath).toBe('coding/personal')
  })

  it('filterTaskTypesByQuery returns all sorted rows for empty query', () => {
    expect(filterTaskTypesByQuery(rows, '').map((t) => t.name)).toEqual([
      'coding',
      'coding/ai',
      'exercise/cardio',
    ])
  })

  it('filterTaskTypesByQuery matches partial root like cod to coding branch', () => {
    expect(filterTaskTypesByQuery(rows, 'cod').map((t) => t.name)).toEqual(['coding', 'coding/ai'])
  })

  it('filterTaskTypesByQuery matches nested partial query coding/a', () => {
    expect(filterTaskTypesByQuery(rows, 'coding/a').map((t) => t.name)).toEqual(['coding', 'coding/ai'])
  })

  it('filterTaskTypesByQuery matches child segment without matching parent root', () => {
    expect(filterTaskTypesByQuery(rows, 'ai').map((t) => t.name)).toEqual(['coding/ai'])
  })

  it('filterTaskTypesByQuery matches query aligned to path suffix segments', () => {
    const deep: TaskType[] = [
      { id: 1, name: 'a/b/c', created_at: '', updated_at: '' },
      { id: 2, name: 'x/y', created_at: '', updated_at: '' },
    ]
    expect(filterTaskTypesByQuery(deep, 'b/c').map((t) => t.name)).toEqual(['a/b/c'])
  })

  it('matches stored names case-insensitively', () => {
    const mixed: TaskType[] = [{ id: 2, name: 'Reading', created_at: '', updated_at: '' }]
    expect(filterTaskTypesByQuery(mixed, 'reading').map((t) => t.name)).toEqual(['Reading'])
    expect(buildTaskTypeSuggestions(mixed, 'Reading').createPath).toBeNull()
  })

  it('filterTaskTypesByQuery matches a repaired double-slash query', () => {
    expect(filterTaskTypesByQuery(rows, 'coding//a').map((t) => t.name)).toEqual([
      'coding',
      'coding/ai',
    ])
  })

  it('taskTypeRootSegment and pathDepth', () => {
    expect(taskTypeRootSegment('coding/ai')).toBe('coding')
    expect(taskTypeRootSegment('work')).toBe('work')
    expect(pathDepth('a')).toBe(0)
    expect(pathDepth('a/b/c')).toBe(2)
  })

  it('groupTaskTypesByRoot sorts and groups by first segment', () => {
    const mixed: TaskType[] = [
      { id: 3, name: 'coding/ai', created_at: '', updated_at: '' },
      { id: 1, name: 'coding', created_at: '', updated_at: '' },
      { id: 2, name: 'dev/x', created_at: '', updated_at: '' },
    ]
    const groups = groupTaskTypesByRoot(mixed)
    expect(groups.map((g) => g.root)).toEqual(['coding', 'dev'])
    expect(groups[0]!.items.map((t) => t.name)).toEqual(['coding', 'coding/ai'])
    expect(groups[1]!.items.map((t) => t.name)).toEqual(['dev/x'])
  })
})

describe('rankTaskTypes', () => {
  const types: TaskType[] = [
    { id: 1, name: 'coding', usage_count: 64, created_at: '', updated_at: '' },
    { id: 2, name: 'coding/ai', usage_count: 41, created_at: '', updated_at: '' },
    { id: 3, name: 'coding/ai/agents', usage_count: 32, created_at: '', updated_at: '' },
    { id: 4, name: 'reading', usage_count: 19, created_at: '', updated_at: '' },
    { id: 9, name: 'unspecified', usage_count: 400, created_at: '', updated_at: '' },
  ]

  it('empty query pins current, then usage, with unspecified last', () => {
    expect(rankTaskTypes(types, '  ', 2).map((t) => t.name)).toEqual([
      'coding/ai',
      'coding',
      'coding/ai/agents',
      'reading',
      'unspecified',
    ])
  })

  it('does not repeat unspecified when it is already current', () => {
    expect(rankTaskTypes(types, '', 9).map((t) => t.name)[0]).toBe('unspecified')
    expect(rankTaskTypes(types, '', 9).filter((t) => t.name === 'unspecified')).toHaveLength(1)
  })

  it('typed query keeps nearby paths for a new slashed path', () => {
    expect(rankTaskTypes(types, 'coding/ai/tooling').map((t) => t.name)).toEqual([
      'coding',
      'coding/ai',
    ])
  })

  it('typed query sorts by path score then name, not usage or current', () => {
    expect(rankTaskTypes(types, 'coding/ai', 1).map((t) => t.name)).toEqual([
      'coding/ai',
      'coding/ai/agents',
      'coding',
    ])
  })
})

describe('createAncestorHint', () => {
  it('names the parent when every ancestor already exists', () => {
    expect(createAncestorHint(rows, 'coding/ai/tooling')).toEqual({
      lead: 'Adds under existing ',
      path: 'coding/ai',
    })
  })

  it('lists only the ancestors that will be created', () => {
    expect(createAncestorHint(rows, 'coding/tooling/scripts')).toEqual({
      lead: 'Also creates ',
      path: 'coding/tooling',
    })
  })

  it('a top level path has nothing to explain', () => {
    expect(createAncestorHint(rows, 'errands')).toBeNull()
  })
})

File: android/app/src/test/java/com/timebox/android/data/TaskTypePathsTest.kt
package com.timebox.android.data

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

class TaskTypePathsTest {

    private val types = listOf(
        taskType(1, "coding", 64),
        taskType(2, "coding/ai", 41),
        taskType(3, "coding/ai/agents", 32),
        taskType(4, "coding/ai/evals", 11),
        taskType(5, "reading", 19),
        taskType(6, "reading/ai-papers", 7),
        taskType(7, "admin", 12),
        taskType(8, "admin/email", 5),
        taskType(9, "meeting", 28),
        taskType(10, "unspecified", 400),
    )

    private fun taskType(id: Int, name: String, usageCount: Int) =
        TaskType(id = id, name = name, usageCount = usageCount)

    private fun names(rows: List<TaskType>) = rows.map { it.name }

    // ---- canonicalization -------------------------------------------------

    @Test
    fun `canonicalize trims and lowercases every segment`() {
        assertEquals("coding/ai/tooling", canonicalizeTaskTypePath("Coding/AI/Tooling "))
    }

    @Test
    fun `canonicalize collapses repeated and trailing slashes`() {
        assertEquals("coding/ai", canonicalizeTaskTypePath("coding//ai/"))
        assertEquals("coding", canonicalizeTaskTypePath("/coding"))
        assertEquals("coding", canonicalizeTaskTypePath("coding/"))
    }

    @Test
    fun `canonicalize collapses internal whitespace`() {
        assertEquals("deep work/writing", canonicalizeTaskTypePath("Deep   Work / Writing"))
    }

    @Test
    fun `canonicalize returns null when nothing survives`() {
        assertNull(canonicalizeTaskTypePath(""))
        assertNull(canonicalizeTaskTypePath("   "))
        assertNull(canonicalizeTaskTypePath("///"))
    }

    @Test
    fun `prefixes list every path the backend touches`() {
        assertEquals(
            listOf("coding", "coding/ai", "coding/ai/agents"),
            taskTypePathPrefixes("coding/ai/agents"),
        )
        assertEquals(listOf("coding"), taskTypePathPrefixes("coding"))
    }

    // ---- two-tone rendering -----------------------------------------------

    @Test
    fun `path parts separate ancestors from the leaf`() {
        assertEquals(
            TaskTypePathParts("coding / ai / ", "agents"),
            taskTypePathParts("coding/ai/agents"),
        )
    }

    @Test
    fun `a single segment is all leaf`() {
        assertEquals(TaskTypePathParts("", "meeting"), taskTypePathParts("meeting"))
    }

    // ---- ranking ----------------------------------------------------------

    @Test
    fun `query ai matches path-aware segments, not email substring`() {
        val ranked = names(rankTaskTypes(types, "ai", currentTypeId = 2))
        assertEquals(
            listOf("coding/ai", "coding/ai/agents", "coding/ai/evals", "reading/ai-papers"),
            ranked,
        )
        assertFalse(ranked.contains("admin/email"))
    }

    @Test
    fun `an exact path outranks a more used prefix match`() {
        val ranked = rankTaskTypes(types, "coding")
        assertEquals("coding", ranked.first().name)
    }

    @Test
    fun `typed query does not pin the current type above a better path match`() {
        val ranked = names(rankTaskTypes(types, "coding/ai", currentTypeId = 1))
        assertEquals("coding/ai", ranked.first())
        assertTrue(ranked.indexOf("coding") > ranked.indexOf("coding/ai/agents"))
    }

    @Test
    fun `a slashed query keeps ancestors and descendants`() {
        assertEquals(
            listOf("coding/ai", "coding/ai/agents", "coding/ai/evals", "coding"),
            names(rankTaskTypes(types, "coding/ai")),
        )
    }

    @Test
    fun `a fully typed new path still shows nearby existing paths`() {
        assertEquals(
            listOf("coding", "coding/ai"),
            names(rankTaskTypes(types, "coding/ai/tooling")),
        )
    }

    @Test
    fun `an empty query pins current, then usage, with unspecified last`() {
        val ranked = names(rankTaskTypes(types, "  ", currentTypeId = 2))
        assertEquals("coding/ai", ranked.first())
        assertEquals("unspecified", ranked.last())
        assertEquals(listOf("coding", "coding/ai/agents", "meeting"), ranked.drop(1).take(3))
    }

    @Test
    fun `unspecified is not repeated when it is already current`() {
        val ranked = names(rankTaskTypes(types, "", currentTypeId = 10))
        assertEquals("unspecified", ranked.first())
        assertEquals(1, ranked.count { it == "unspecified" })
    }

    // ---- create affordance ------------------------------------------------

    @Test
    fun `create is offered only for a path that does not exist`() {
        assertTrue(shouldOfferCreate(types, "coding/ai/tooling"))
        assertTrue(shouldOfferCreate(types, "Coding/AI/Tooling "))
        assertTrue(shouldOfferCreate(types, "coding//ai/tooling/"))
        assertFalse(shouldOfferCreate(types, "coding/ai"))
        assertFalse(shouldOfferCreate(types, "  CODING / AI  "))
        assertFalse(shouldOfferCreate(types, ""))
    }

    @Test
    fun `hint names the parent when every ancestor already exists`() {
        assertEquals(
            CreateAncestorHint("Adds under existing ", "coding/ai"),
            createAncestorHint(types, "coding/ai/tooling"),
        )
    }

    @Test
    fun `hint lists only the ancestors that will be created`() {
        assertEquals(
            CreateAncestorHint("Also creates ", "coding/tooling"),
            createAncestorHint(types, "coding/tooling/scripts"),
        )
    }

    @Test
    fun `a top level path has nothing to explain`() {
        assertNull(createAncestorHint(types, "errands"))
    }
}
